src/NServiceBus.Core/Unicast/Behaviors/MessageHandler.cs

```diff
@@ -7,14 +7,75 @@
     /// </summary>
     public class MessageHandler
     {
+        Action<object, object, object> invocationWithContext;
+
+        /// <summary>
+        /// Creates a new instance of the message handler
+        /// </summary>
+        [ObsoleteEx(TreatAsErrorFromVersion = "6", RemoveInVersion = "7", ReplacementTypeOrMember = "MessageHandler(Action<object, object, object> invocationWithContext, Type handlerType)")]
+        public MessageHandler()
+        {
+            throw new NotImplementedException("Creator of the message handler must assign the handler type and the invocation delegate");
+        }
+
         /// <summary>
-        /// The actual instance, can be a saga or just a plain handler
+        /// Creates a new instance of the message handler with predefined invocation delegate and handler type
+        /// </summary>
+        /// <param name="invocationWithContext">The invocation with context delegate</param>
+        /// <param name="handlerType">The handler type</param>
+        /// <param name="handlerKind">The handler kind</param>
+        // TODO: Should we expose that to public?
+        internal MessageHandler(Action<object, object, object> invocationWithContext, Type handlerType, HandlerKind handlerKind)
+        {
+            HandlerKind = handlerKind;
+            HandlerType = handlerType;
+            this.invocationWithContext = invocationWithContext;
+        }
+
+        /// <summary>
+        /// The actual instance, can be a saga, a timeout or just a plain handler
         /// </summary>
         public object Instance { get; set; }
-        
+
+        /// <summary>
+        /// The handler type, can be a saga, a timeout or just a plain handler
+        /// </summary>
+        public Type HandlerType { get; private set; }
+
+        /// <summary>
+        /// TODO: Should we expose that to public?
+        /// </summary>
+        internal HandlerKind HandlerKind { get; private set; }
+
         /// <summary>
         /// The actual invocation
         /// </summary>
-        public Action<object, object> Invocation { get; set; }
+        [ObsoleteEx(TreatAsErrorFromVersion = "6", RemoveInVersion = "7", ReplacementTypeOrMember = "MessageHandler.Invoke")]
+        public Action<object, object> Invocation
+        {
+            get { throw new NotImplementedException(); }
+            set { throw new NotImplementedException(); }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="message"></param>
+        /// <param name="context"></param>
+        public void Invoke(object message, object context)
+        {
+            invocationWithContext(Instance, message, context);
+        }
+    }
+
+    /// <summary>
+    /// TODO: Should we expose that to public?
+    /// </summary>
+    internal enum HandlerKind
```

Comment:
moving this to its own file will make it more discoverable in the solution exploirer